machinist/node.go

```diff
@@ -0,0 +1,161 @@
+package machinist
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"github.com/enfabrica/enkit/lib/client/ccontext"
+	"github.com/enfabrica/enkit/machinist/rpc/machinist"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"time"
+)
+
+type NodeErr struct {
+}
+type Node struct {
+	Client      machinist.ControllerClient
+	TLSCert     string
+	Connection  *grpc.ClientConn
+	KillChannel chan NodeErr
+	PongChannel chan *machinist.PollResponse
+	*ccontext.Context
+}
+
+func NewNode(nm ...NodeModifier) (*Node, error) {
+	n := &Node{
+		KillChannel: make(chan NodeErr),
+		Context:     ccontext.DefaultContext(),
+	}
+	for _, mod := range nm {
+		if err := mod(n); err != nil {
+			fmt.Println("there was an error", err.Error())
+			return nil, err
+		}
+	}
+	return n, nil
+}
+
+type NodeModifier func(node *Node) error
+
+func (n *Node) listenForPong() {
+	ctx := context.Background()
+	pollStream, err := n.Client.Poll(ctx)
+	if err != nil {
+		n.Kill(err)
+	}
+	for {
+		pollResponse, err := pollStream.Recv()
+		if err != nil {
+			n.Kill(err)
+		}
+		n.PongChannel <- pollResponse
+	}
+
+}
+
+func (n *Node) BeginPolling() {
+	ctx := context.Background()
+	pollStream, err := n.Client.Poll(ctx)
+	if err != nil {
+		panic(err)
+	}
+	prometheusPayload, err := n.GatherInformation()
+	if err != nil {
+		n.Kill(err)
+	}
+	err = pollStream.Send(&machinist.PollRequest{
+		Req: &machinist.PollRequest_Ping{
+			Ping: &machinist.ClientPing{
+				Payload: prometheusPayload,
+			},
+		},
+	})
+	if err != nil {
+		n.Kill(err)
+	}
+	fmt.Println("in here")
+	for {
+		select {
+		case nodeErr := <-n.KillChannel:
+			fmt.Printf("killing node with error %v", nodeErr)
+			return
+		//TODO(adam): include symmetic encodeer token as extra layer of security in case node is comped
+		case pollResp := <-n.PongChannel:
+			fmt.Println("recv %v", pollResp)
+			// TODO(adam): should scheduling happen here or should it happen somewhere else
+		case <-time.After(2 * time.Second):
+			prometheusPayload, err := n.GatherInformation()
+			if err != nil {
+				n.Kill(err)
+			}
+			fmt.Println("sending")
+			err = pollStream.Send(&machinist.PollRequest{
+				Req: &machinist.PollRequest_Ping{
+					Ping: &machinist.ClientPing{
+						Payload: prometheusPayload,
+					},
+				},
+			})
+			fmt.Println("sent")
+			if err != nil {
+				n.Kill(err)
+			}
+		}
+	}
+}
+
+func (n *Node) Kill(err error) {
+	fmt.Println("killing with err", err.Error())
+	n.KillChannel <- NodeErr{}
+}
+
+func (n Node) Stop() {
+	n.KillChannel <- NodeErr{}
+}
+
+func WithInviteToken(token string) NodeModifier {
+	return func(node *Node) error {
+		data, err := base64.RawStdEncoding.DecodeString(token)
```

Comment:
when we last chat, I was expecting the invitation token to be a short string? this looks like it'll be a large blob of data?